File: examples/numerical-models/python/src/host_scale.py
#!/usr/bin/env python3
"""Host-scale oracle JSON artifacts under results/<slug>/."""

from __future__ import annotations

import json
from pathlib import Path

import numpy as np
from numpy.polynomial.polynomial import polyfromroots, polyval
from scipy import linalg, signal
from scipy.interpolate import RegularGridInterpolator

CRATE_ROOT = Path(__file__).resolve().parents[2]
TAU = 10.0
EPS = float(np.finfo(np.float64).eps)


def save(path: Path, doc: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(doc, indent=2) + "\n", encoding="utf-8")
    print(f"wrote {path}")


def _eval_kappa(coeffs: np.ndarray, x: float) -> float:
    ax = abs(x)
    tilde = float(np.polynomial.polynomial.polyval(ax, np.abs(coeffs)))
    val = abs(float(np.polynomial.polynomial.polyval(x, coeffs)))
    if val == 0.0:
        return float("inf")
    return tilde / val


def _well_conditioned(n: int, scale: float) -> np.ndarray:
    idx = np.arange(n, dtype=np.float64)
    return np.eye(n) + scale / (idx[:, None] + idx[None, :] + 1.0)


def emit_hilbert(n: int) -> None:
    a = linalg.hilbert(n)
    b = np.ones(n, dtype=np.float64)
    x = linalg.solve(a, b)
    kappa = float(np.linalg.cond(a))
    save(
        CRATE_ROOT / "results" / "matrix" / f"host_hilbert{n}_python.json",
        {
            "slug": "matrix",
            "source": "python",
            "fixture": f"hilbert{n}",
            "values": {
                "A": a.tolist(),
                "b": b.reshape(n, 1).tolist(),
                "x_gold": x.reshape(n, 1).tolist(),
                "kappa": kappa,
                "keps": kappa * EPS,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )


def emit_gemm_lu1024() -> None:
    a = _well_conditioned(1024, 0.01)
    b = _well_conditioned(1024, 0.02)
    c = a @ b
    rhs = np.ones(1024, dtype=np.float64)
    x = linalg.solve(a, rhs)
    kappa = float(np.linalg.cond(a))
    save(
        CRATE_ROOT / "results" / "matrix" / "host_gemm1024_python.json",
        {
            "slug": "matrix",
            "source": "python",
            "fixture": "gemm1024",
            "values": {
                "C_gold": c.tolist(),
                "kappa": kappa,
                "keps": kappa * EPS,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )
    save(
        CRATE_ROOT / "results" / "matrix" / "host_lu1024_python.json",
        {
            "slug": "matrix",
            "source": "python",
            "fixture": "lu1024",
            "values": {
                "b": rhs.reshape(1024, 1).tolist(),
                "x_gold": x.reshape(1024, 1).tolist(),
                "kappa": kappa,
                "keps": kappa * EPS,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )


def emit_poly52() -> None:
    roots = np.arange(1.0, 52.0, dtype=np.float64)
    coeffs = np.asarray(polyfromroots(roots), dtype=np.float64)
    poly_x = 20.5
    poly_val = float(polyval(poly_x, coeffs))
    poly_kappa = _eval_kappa(coeffs, poly_x)
    save(
        CRATE_ROOT / "results" / "polynomial" / "host_poly52_python.json",
        {
            "slug": "polynomial",
            "source": "python",
            "fixture": "poly52",
            "values": {
                "coeffs": coeffs.tolist(),
                "poly_x": poly_x,
                "poly_val": poly_val,
                "kappa": poly_kappa,
                "keps": poly_kappa * EPS,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )


def emit_tf_clustered() -> None:
    roots = np.arange(1.0, 52.0, dtype=np.float64)
    coeffs = np.asarray(polyfromroots(roots), dtype=np.float64)
    tf_num = np.array([1.0], dtype=np.float64)
    tf_w = np.array([0.5, 1.0, 2.0], dtype=np.float64)
    _w, h = signal.freqs(np.flip(tf_num), np.flip(coeffs), worN=tf_w)
    tf_kappa = max(_eval_kappa(coeffs, float(w)) for w in tf_w)
    save(
        CRATE_ROOT / "results" / "transfer_function" / "host_tf_clustered_python.json",
        {
            "slug": "transfer_function",
            "source": "python",
            "fixture": "tf_clustered",
            "values": {
                "den": coeffs.tolist(),
                "omegas": tf_w.tolist(),
                "h_re": np.asarray(h.real, dtype=np.float64).tolist(),
                "h_im": np.asarray(h.imag, dtype=np.float64).tolist(),
                "kappa": tf_kappa,
                "keps": tf_kappa * EPS,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )


def emit_stiff_zoh() -> None:
    a_stiff = np.diag([-1.0, -1.0e6])
    tmat = np.array([[1.0, 1.0], [0.0, 1.0]], dtype=np.float64)
    a_scr = tmat @ a_stiff @ np.linalg.inv(tmat)
    b_c = np.array([[0.0], [1.0]], dtype=np.float64)
    c_c = np.array([[1.0, 0.0]], dtype=np.float64)
    d_c = np.array([[0.0]], dtype=np.float64)
    dt = 1.0e-6
    ad, _bd, _cd, _dd, _ = signal.cont2discrete(
        (a_scr, b_c, c_c, d_c), dt, method="zoh"
    )
    stiff_kappa = float(np.linalg.cond(a_scr))
    save(
        CRATE_ROOT / "results" / "state_space" / "host_stiff_zoh_python.json",
        {
            "slug": "state_space",
            "source": "python",
            "fixture": "stiff_zoh",
            "values": {
                "A": a_scr.tolist(),
                "Ad_gold": np.asarray(ad, dtype=np.float64).tolist(),
                "dt": dt,
                "kappa": stiff_kappa,
                "tau": TAU,
                "eps": EPS,
            },
            "series": {},
        },
    )


def emit_tensor1024() -> None:
    n_g = 1024
    idx = np.arange(n_g, dtype=np.float32)
    grid = idx[:, None] + idx[None, :]
    axes = (np.arange(n_g, dtype=np.float32), np.arange(n_g, dtype=np.float32))
    interp = RegularGridInterpolator(
        axes, grid, method="linear", bounds_error=False, fill_value=None
    )
    points = np.array(
        [[0.0, 0.0], [10.5, 20.25], [100.0, 200.0], [512.5, 512.5]],
        dtype=np.float32,
    )
    samples = np.asarray(interp(points), dtype=np.float32)
    save(
        CRATE_ROOT / "results" / "tensor" / "host_tensor1024_python.json",
        {
            "slug": "tensor",
            "source": "python",
            "fixture": "tensor1024",
            "values": {
                "points": points.tolist(),
                "samples_gold": samples.tolist(),
            },
            "series": {},
        },
    )


def main() -> None:
    for n in (12, 16, 128):
        emit_hilbert(n)
    emit_gemm_lu1024()
    emit_poly52()
    emit_tf_clustered()
    emit_stiff_zoh()
    emit_tensor1024()


if __name__ == "__main__":
    main()

File: examples/numerical-models/src/host_scale.rs
//! Host-scale JSON V&V — reads `results/<slug>/host_*_{python,native}.json`.

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use serde_json::Value;

    use crate::SOLVE_RESIDUAL_TAU;

    const RESIDUAL_TAU: f64 = SOLVE_RESIDUAL_TAU;

    fn manifest_dir() -> PathBuf {
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
    }

    fn load_artifact(rel: &str, hint: &str) -> Value {
        let path = manifest_dir().join(rel);
        let text = std::fs::read_to_string(&path).unwrap_or_else(|_| {
            panic!("missing artifact: {}\nrun: {}", path.display(), hint);
        });
        serde_json::from_str(&text).expect("parse json")
    }

    fn report_only_bound_miss(claim: &str, detail: &str) {
        println!("report-only host-scale bound: {claim} — {detail}");
    }

    fn assert_hilbert(n: usize, gate: bool) {
        let py = load_artifact(
            &format!("results/matrix/host_hilbert{n}_python.json"),
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            &format!("results/matrix/host_hilbert{n}_native.json"),
            "cargo run --example host_scale_emit",
        );
        let meta = &py["values"];
        let got = &rs["values"];
        let kappa = meta["kappa"].as_f64().unwrap();
        let keps = meta["keps"].as_f64().unwrap();
        let tau = meta["tau"].as_f64().unwrap();
        let residual_ratio = got["residual_ratio"].as_f64().unwrap();
        let rel_err = got["rel_err"].as_f64().unwrap();
        println!("hilbert{n} residual ratio={residual_ratio:.6e} kappa={kappa:.6e}");
        if gate && keps < 1.0 {
            assert!(residual_ratio < RESIDUAL_TAU, "hilbert{n} residual {residual_ratio}");
            let bound = tau * keps;
            assert!(rel_err <= bound, "hilbert{n} forward rel {rel_err} > {bound}");
        } else {
            println!("hilbert{n} κε≳1 or report-only: rel={rel_err:.6e}");
        }
    }

    #[test]
    fn host_scale_hilbert() {
        assert_hilbert(12, true);
        assert_hilbert(16, true);
        assert_hilbert(128, false);
    }

    #[test]
    fn host_scale_poly_clustered() {
        let py = load_artifact(
            "results/polynomial/host_poly52_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/polynomial/host_poly52_native.json",
            "cargo run --example host_scale_emit",
        );
        let meta = &py["values"];
        let rel = rs["values"]["rel_err"].as_f64().unwrap();
        let bound = meta["tau"].as_f64().unwrap() * meta["kappa"].as_f64().unwrap() * meta["eps"].as_f64().unwrap();
        println!("poly52 rel={rel:.6e} bound={bound:.6e}");
        if meta["kappa"].as_f64().unwrap() * meta["eps"].as_f64().unwrap() < 1.0 {
            assert!(rel <= bound, "poly52 rel {rel} > {bound}");
        }
    }

    #[test]
    fn host_scale_tf_clustered() {
        let py = load_artifact(
            "results/transfer_function/host_tf_clustered_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/transfer_function/host_tf_clustered_native.json",
            "cargo run --example host_scale_emit",
        );
        let meta = &py["values"];
        let rels = rs["values"]["rel_errs"].as_array().unwrap();
        let bound = meta["tau"].as_f64().unwrap() * meta["kappa"].as_f64().unwrap() * meta["eps"].as_f64().unwrap();
        for (i, rel) in rels.iter().enumerate() {
            let rel = rel.as_f64().unwrap();
            println!("tf w idx={i} rel={rel:.6e} bound={bound:.6e}");
            if meta["kappa"].as_f64().unwrap() * meta["eps"].as_f64().unwrap() < 1.0 {
                assert!(rel <= bound, "tf rel {rel} > {bound}");
            }
        }
    }

    #[test]
    fn host_scale_stiff_zoh() {
        let _py = load_artifact(
            "results/state_space/host_stiff_zoh_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/state_space/host_stiff_zoh_native.json",
            "cargo run --example host_scale_emit",
        );
        let _rel = rs["values"]["rel_err"].as_f64().unwrap();
        let ratio = rs["values"]["residual_ratio"].as_f64().unwrap();
        println!("stiff ZOH residual ratio={ratio:.6e}");
        if ratio >= RESIDUAL_TAU {
            report_only_bound_miss(
                "stiff ZOH A_d residual ratio",
                &format!("ratio={ratio:.6e} (threshold {RESIDUAL_TAU})"),
            );
        }
    }

    #[test]
    #[ignore = "1024x1024 tensor grid; slow. Run: cargo test -- --ignored"]
    fn host_scale_tensor_grid() {
        let py = load_artifact(
            "results/tensor/host_tensor1024_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/tensor/host_tensor1024_native.json",
            "cargo run --example host_scale_emit",
        );
        let errs = rs["values"]["abs_errs"].as_array().unwrap();
        let points = py["values"]["points"].as_array().unwrap();
        for (pt, err) in points.iter().zip(errs.iter()) {
            let err = err.as_f64().unwrap() as f32;
            println!("tensor1024 {:?} abs={err}", pt);
            assert!(err <= 1e-3, "tensor interp {err}");
        }
    }

    #[test]
    #[ignore = "1024x1024 GEMM; slow. Run: cargo test -- --ignored"]
    fn host_scale_1024_gemm() {
        let py = load_artifact(
            "results/matrix/host_gemm1024_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/matrix/host_gemm1024_native.json",
            "cargo run --example host_scale_emit",
        );
        let rel = rs["values"]["rel_err"].as_f64().unwrap();
        let bound = py["values"]["tau"].as_f64().unwrap()
            * py["values"]["kappa"].as_f64().unwrap()
            * py["values"]["eps"].as_f64().unwrap();
        println!("gemm1024 rel={rel:.6e} bound={bound:.6e}");
        if rel > bound {
            report_only_bound_miss(
                "1024x1024 GEMM forward error",
                &format!("rel={rel:.6e} bound={bound:.6e}"),
            );
        }
    }

    #[test]
    #[ignore = "1024x1024 LU; slow. Run: cargo test -- --ignored"]
    fn host_scale_1024_lu() {
        let py = load_artifact(
            "results/matrix/host_lu1024_python.json",
            "python3 python/src/host_scale.py",
        );
        let rs = load_artifact(
            "results/matrix/host_lu1024_native.json",
            "cargo run --example host_scale_emit",
        );
        let residual_ratio = rs["values"]["residual_ratio"].as_f64().unwrap();
        let rel_x = rs["values"]["rel_err"].as_f64().unwrap();
        let bound = py["values"]["tau"].as_f64().unwrap()
            * py["values"]["kappa"].as_f64().unwrap()
            * py["values"]["eps"].as_f64().unwrap();
        println!("lu1024 residual={residual_ratio:.6e} rel={rel_x:.6e} bound={bound:.6e}");
        if residual_ratio >= RESIDUAL_TAU {
            report_only_bound_miss(
                "1024x1024 LU residual ratio",
                &format!("residual={residual_ratio:.6e} (threshold {RESIDUAL_TAU})"),
            );
        }
        if rel_x > bound {
            report_only_bound_miss(
                "1024x1024 LU forward error",
                &format!("rel={rel_x:.6e} bound={bound:.6e}"),
            );
        }
    }
}
